src/topology/instances/add_circle.lean

```diff
@@ -286,3 +286,115 @@ local attribute [instance] fact_zero_lt_one
 /-- The unit circle `ℝ ⧸ ℤ`. -/
 @[derive [compact_space, normal_space, second_countable_topology]]
 abbreviation unit_add_circle := add_circle (1 : ℝ)
+
+
+section identify_Icc_ends
+/-! This section proves that the natural map from `[0, T] ⊂ ℝ` to `add_circle T` gives an
+identification of `add_circle T`, as a topological space, with the quotient of `[0, T]` by the
+equivalence relation identifying `0` with `T`. -/
+
+variables (T : ℝ) [hT : fact (0 < T)]
+include hT
+local notation `𝕋` := add_circle T
+
+private lemma Icc_to_circle_kernel {x y : ℝ} (hx : x ∈ Icc 0 T) (hy : y ∈ Icc 0 T) :
+  (x : 𝕋) = (y : 𝕋) ↔ (x = y) ∨ (x = 0 ∧ y = T) ∨ (y = 0 ∧ x = T) :=
+begin
+  simp_rw [quotient_add_group.eq, add_subgroup.mem_zmultiples_iff, zsmul_eq_mul],
+  split,
+  { rintros ⟨n, hn⟩,
+    have : | -x + y| ≤ T,
+    { rw abs_le,
+      exact ⟨by linarith [hx.2, hy.1], by linarith [hx.1, hy.2]⟩, },
+    rw [←hn, abs_mul, abs_of_pos hT.out, mul_le_iff_le_one_left hT.out, ←int.cast_abs,
+      ←int.cast_one, int.cast_le] at this,
+    replace this : |n| = 0 ∨ |n| = 1,
+    { rcases eq_or_lt_of_le this with h|h,
+      { exact or.inr h, },
+      { exact or.inl (by linarith [abs_nonneg n]) } },
+    rw [abs_eq (zero_le_one' ℤ), abs_eq_zero] at this,
+    rcases this with (rfl | rfl | rfl),
+    { rw int.cast_zero at hn,
+      exact or.inl (by linarith), },
+    { rw int.cast_one at hn,
+      exact or.inr (or.inl ⟨by linarith [hx.1, hy.2], by linarith [hx.1, hy.2]⟩) },
+    { rw [int.cast_neg, int.cast_one] at hn,
+      exact or.inr (or.inr ⟨by linarith [hy.1, hx.2], by linarith [hy.1, hx.2]⟩), }, },
+  { rintros (rfl | ⟨rfl, h⟩ | ⟨rfl, h⟩),
+    exacts [ ⟨0, by simp⟩, ⟨1, by {rw h, simp}⟩, ⟨-1, by {rw h, simp}⟩] },
+end
+
+variables {T}
+
+/-- The equivalence relation on `Icc 0 T` which identifies `0` and `T`. -/
+private def S : setoid (Icc 0 T) :=
+{ r := λ x y, (x.1 = y.1) ∨ (x.1 = 0 ∧ y.1 = T) ∨ (y.1 = 0 ∧ x.1 = T),
+  iseqv := ⟨(λ x, by tauto), (λ x y hxy, by tauto),
+    (λ x y z hxy hyz, (Icc_to_circle_kernel T x.2 z.2).mp
+    (((Icc_to_circle_kernel T x.2 y.2).mpr hxy).trans
+    ((Icc_to_circle_kernel T y.2 z.2).mpr hyz)))⟩ }
+
+local attribute [instance] S
```

Comment:
Instead of making this a local instance, it would probably be a better ideal to use the primed `quotient.lift_on'` etc definitions instead; that's how we handle the quotient on `add_circle`.